homeassistant/components/otbr/homeassistant_hardware.py

```diff
@@ -0,0 +1,80 @@
+"""Home Assistant Hardware firmware utilities."""
+
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING, cast
+
+from yarl import URL
+
+from homeassistant.components.hassio import AddonManager
+from homeassistant.components.homeassistant_hardware.util import (
+    ApplicationType,
+    FirmwareInfo,
+    OwningAddon,
+    OwningIntegration,
+    get_otbr_addon_firmware_info,
+)
+from homeassistant.config_entries import ConfigEntry, ConfigEntryState
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.hassio import is_hassio
+
+from .const import DOMAIN
+
+if TYPE_CHECKING:
+    from . import OTBRConfigEntry
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_get_firmware_info(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> FirmwareInfo | None:
+    """Return firmware information for the OpenThread Border Router."""
+    owners: list[OwningIntegration | OwningAddon] = [
+        OwningIntegration(config_entry_id=config_entry.entry_id)
+    ]
+
+    device = None
+
+    if is_hassio(hass) and (host := URL(config_entry.data["url"]).host) is not None:
+        otbr_addon_manager = AddonManager(
+            hass=hass,
+            logger=_LOGGER,
+            addon_name="OpenThread Border Router",
+            addon_slug=host,
+        )
+
+        if (
+            addon_fw_info := await get_otbr_addon_firmware_info(
+                hass, otbr_addon_manager
+            )
+        ) is not None:
+            device = addon_fw_info.device
+            owners.extend(addon_fw_info.owners)
+
+    firmware_version = None
+
+    if config_entry.state in (
+        ConfigEntryState.LOADED,
+        ConfigEntryState.SETUP_IN_PROGRESS,
```

Comment:
This config entry state is required because we call `async_get_firmware_info` during OTBR setup.

Reply:
Turned into a comment: https://github.com/home-assistant/core/pull/138330/commits/033b215be6de62bb6d9ae6e7d68f0e20f41b6ed0